File: src/main/java/com/example/finsentinel/service/scraper/PolygonNewsScraper.java
package com.example.finsentinel.service.scraper;

import com.example.finsentinel.config.PolygonProperties;
import com.example.finsentinel.model.Document;
import com.example.finsentinel.model.enums.DocumentType;
import com.example.finsentinel.repository.DocumentRepository;
import com.example.finsentinel.service.storage.MinioStorageService;
import com.example.finsentinel.util.MarkdownToPdfConverter;
import com.example.finsentinel.stream.VectorizeStreamProducer;
import com.example.finsentinel.util.SectorMapper;
import com.fasterxml.jackson.databind.JsonNode;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.UUID;

@Service
@Slf4j
@RequiredArgsConstructor
public class PolygonNewsScraper {

    private final PolygonProperties polygonProperties;
    private final MinioStorageService storageService;
    private final DocumentRepository documentRepository;
    private final RestClient restClient;
    private final VectorizeStreamProducer vectorizeStreamProducer;
    private final FirecrawlClient firecrawlClient;

    /**
     * Fetch financial news from Polygon.io API for the given tickers.
     * Uses Firecrawl to scrape full article content when available.
     *
     * @param tickers list of stock tickers
     * @param days    how many days back to fetch
     * @return number of news articles saved
     */
    public int scrape(List<String> tickers, int days) {
        log.info("Starting Polygon news scrape for tickers: {}, days: {}", tickers, days);
        int successCount = 0;

        String publishedAfter = LocalDate.now().minusDays(days)
                .format(DateTimeFormatter.ISO_LOCAL_DATE);

        for (String ticker : tickers) {
            try {
                JsonNode response = restClient.get()
                        .uri(polygonProperties.getBaseUrl() +
                                        "/v2/reference/news?ticker={ticker}&published_utc.gte={date}&limit=50&apiKey={apiKey}",
                                ticker, publishedAfter, polygonProperties.getApiKey())
                        .retrieve()
                        .body(JsonNode.class);

                if (response != null && response.has("results")) {
                    for (JsonNode article : response.get("results")) {
                        try {
                            if (saveArticle(article, ticker)) {
                                successCount++;
                            }
                        } catch (Exception e) {
                            log.error("Failed to save news article for {}", ticker, e);
                        }
                    }
                }
            } catch (Exception e) {
                log.error("Failed to fetch Polygon news for {}", ticker, e);
            }
        }

        log.info("Polygon news scrape complete: {} articles saved", successCount);
        return successCount;
    }

    private boolean saveArticle(JsonNode article, String ticker) {
        String title = article.has("title") ? article.get("title").asText() : "Untitled";
        String description = article.has("description") ? article.get("description").asText() : "";
        String author = article.has("author") ? article.get("author").asText() : "Unknown";
        String publishedUtc = article.has("published_utc") ? article.get("published_utc").asText() : "";
        String articleUrl = article.has("article_url") ? article.get("article_url").asText() : "";

        // Dedup: skip if already scraped
        if (documentRepository.existsByOriginalFileName(title)) {
            log.debug("Skipping duplicate news article: {}", title);
            return false;
        }

        // Try to scrape full article content via Firecrawl, fall back to description
        String fullContent = description;
        if (!articleUrl.isEmpty()) {
            try {
                FirecrawlClient.ScrapeResult scraped = firecrawlClient.scrape(articleUrl);
                if (scraped != null && !scraped.markdown().isBlank()) {
                    fullContent = scraped.markdown();
                    log.debug("Scraped full article for: {}", title);
                }
            } catch (Exception e) {
                log.warn("Firecrawl failed for {}, using description fallback", articleUrl);
            }
        }

        // Build markdown content
        StringBuilder markdown = new StringBuilder();
        markdown.append("# ").append(title).append("\n\n");
        markdown.append("**Author:** ").append(author).append("\n");
        markdown.append("**Published:** ").append(publishedUtc).append("\n");
        markdown.append("**Ticker:** ").append(ticker).append("\n");
        if (!articleUrl.isEmpty()) {
            markdown.append("**Source:** ").append(articleUrl).append("\n");
        }
        markdown.append("\n---\n\n");
        markdown.append(fullContent);

        // Include keywords/tickers if available
        if (article.has("tickers")) {
            markdown.append("\n\n**Related Tickers:** ");
            for (JsonNode t : article.get("tickers")) {
                markdown.append(t.asText()).append(" ");
            }
        }

        byte[] pdfBytes = MarkdownToPdfConverter.convert(title, markdown.toString());

        String safeTitle = title.replaceAll("[^a-zA-Z0-9\\s-]", "").trim();
        if (safeTitle.length() > 80) safeTitle = safeTitle.substring(0, 80);

        String storageKey = "news/" + ticker + "/" + UUID.randomUUID() + ".pdf";
        storageService.upload(storageKey, pdfBytes, "application/pdf");

        Document doc = Document.builder()
                .fileName(safeTitle + ".pdf")
                .originalFileName(title)
                .docType(DocumentType.NEWS)
                .sector(SectorMapper.fromTicker(ticker))
                .regionId("US")
                .fileSize((long) pdfBytes.length)
                .storageKey(storageKey)
                .build();
        documentRepository.save(doc);
        vectorizeStreamProducer.send(doc.getId());

        log.debug("Saved news article: {} for {}", title, ticker);
        return true;
    }

}

File: apps/api/src/scraper/polygon-news.scraper.ts
import { Injectable, Inject, Logger, Optional } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { documents, eq } from '@finsentinel/db';
import type { DrizzleDB } from '@finsentinel/db';
import { FirecrawlClient } from './firecrawl.client';
import { VectorizeProducer } from '../queue/vectorize.producer';

interface PolygonNewsResponse {
  results: Array<{
    id: string;
    title: string;
    article_url: string;
    author: string;
    published_utc: string;
    description?: string;
    tickers?: string[];
  }>;
  status: string;
  count: number;
}

/**
 * Polygon.io news scraper — fetches news articles from Polygon API.
 *
 * For each article, scrapes full content via Firecrawl if URL is available,
 * saves to Document table, and queues for vectorization.
 */
@Injectable()
export class PolygonNewsScraper {
  private readonly logger = new Logger(PolygonNewsScraper.name);
  private readonly apiKey: string;

  private static readonly BASE_URL = 'https://api.polygon.io/v2/reference/news';

  constructor(
    @Inject('DRIZZLE_DB') private readonly db: DrizzleDB,
    private readonly firecrawl: FirecrawlClient,
    configService: ConfigService,
    @Optional() private readonly vectorizeProducer?: VectorizeProducer,
  ) {
    this.apiKey = configService.get<string>('polygon.apiKey', '');
  }

  /**
   * Scrape news articles for the given tickers.
   * Returns the count of newly saved documents.
   */
  async scrape(tickers: string[]): Promise<number> {
    let savedCount = 0;

    for (const ticker of tickers) {
      try {
        const count = await this.scrapeForTicker(ticker);
        savedCount += count;
      } catch (err) {
        this.logger.error(
          `Polygon news scrape failed for ${ticker}: ${err instanceof Error ? err.message : String(err)}`,
        );
      }
    }

    return savedCount;
  }

  private async scrapeForTicker(ticker: string): Promise<number> {
    const url = new URL(PolygonNewsScraper.BASE_URL);
    url.searchParams.set('ticker', ticker);
    url.searchParams.set('limit', '10');
    url.searchParams.set('apiKey', this.apiKey);

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Polygon API returned ${response.status}`);
    }

    const data = (await response.json()) as PolygonNewsResponse;
    const articles = data.results ?? [];

    this.logger.log(`Polygon found ${articles.length} news articles for ${ticker}`);

    let savedCount = 0;

    for (const article of articles) {
      const originalFileName = `polygon-news-${article.id}`;

      // Dedup: skip if document already exists
      const existing = await this.db
        .select({ id: documents.id })
        .from(documents)
        .where(eq(documents.originalFileName, originalFileName))
        .limit(1);

      if (existing.length > 0) {
        continue;
      }

      if (!article.article_url) {
        continue;
      }

      try {
        const markdown = await this.firecrawl.scrape(article.article_url);
        if (!markdown) {
          this.logger.warn(`No content scraped for article ${article.id}`);
          continue;
        }

        const [inserted] = await this.db
          .insert(documents)
          .values({
            fileName: `${ticker}-news-${article.id}.md`,
            originalFileName,
            docType: 'NEWS_ARTICLE',
            status: 'PENDING',
            sector: null,
            regionId: 'US',
            fileSize: Buffer.byteLength(markdown, 'utf-8'),
          })
          .returning({ id: documents.id });

        if (inserted?.id && this.vectorizeProducer) {
          await this.vectorizeProducer.send(inserted.id);
          this.logger.log(`Enqueued vectorization for doc ${inserted.id}`);
        }
        savedCount++;
      } catch (err) {
        this.logger.warn(
          `Failed to scrape article ${article.id}: ${err instanceof Error ? err.message : String(err)}`,
        );
      }
    }

    return savedCount;
  }
}
